File: go/scheduler.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
)

type Employee struct {
	Name  string            `json:"name"`
	Prefs map[string]string `json:"prefs"`
}
type Config struct {
	Employees   []Employee `json:"employees"`
	Requirement int        `json:"requirement"`
	MaxDays     int        `json:"max_days"`
	Shifts      []string   `json:"shifts"`
	Days        []string   `json:"days"`
}

func main() {
	rand.Seed(7)
	b, err := ioutil.ReadFile("../data/sample_input.json")
	if err != nil {
		fmt.Println("read error:", err)
		return
	}
	var cfg Config
	if err := json.Unmarshal(b, &cfg); err != nil {
		fmt.Println("json error:", err)
		return
	}

	// build structures
	sched := map[string]map[string][]string{}
	for _, d := range cfg.Days {
		sched[d] = map[string][]string{}
		for _, sh := range cfg.Shifts {
			sched[d][sh] = []string{}
		}
	}
	worked := map[string]int{}
	assigned := map[string]map[string]bool{}
	for _, e := range cfg.Employees {
		worked[e.Name] = 0
		assigned[e.Name] = map[string]bool{}
	}

	// pass 1: honor preferences
	for _, d := range cfg.Days {
		for _, e := range cfg.Employees {
			if pref, ok := e.Prefs[d]; ok {
				if len(sched[d][pref]) < cfg.Requirement && worked[e.Name] < cfg.MaxDays && !assigned[e.Name][d] {
					sched[d][pref] = append(sched[d][pref], e.Name)
					assigned[e.Name][d] = true
					worked[e.Name]++
				}
			}
		}
	}

	// pass 2: fill minimums with progress check to avoid infinite loop
	for _, d := range cfg.Days {
		for _, sh := range cfg.Shifts {
			progress := true
			for len(sched[d][sh]) < cfg.Requirement && progress {
				progress = false
				for _, e := range cfg.Employees {
					if worked[e.Name] < cfg.MaxDays && !assigned[e.Name][d] {
						sched[d][sh] = append(sched[d][sh], e.Name)
						assigned[e.Name][d] = true
						worked[e.Name]++
						progress = true
						if len(sched[d][sh]) >= cfg.Requirement {
							break
						}
					}
				}
			}
			// if still short, leave it short
		}
	}

	// print
	for _, d := range cfg.Days {
		fmt.Println(d)
		for _, sh := range cfg.Shifts {
			crew := "(none)"
			if len(sched[d][sh]) > 0 {
				crew = ""
				for i, n := range sched[d][sh] {
					if i > 0 {
						crew += ", "
					}
					crew += n
				}
			}
			fmt.Printf("  %s: %s\n", sh, crew)
		}
		fmt.Println()
	}
}

File: python/scheduler.py
import json, random
from collections import defaultdict

random.seed(7)

def load(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)

def main():
    cfg = load("data/sample_input.json")
    employees = cfg["employees"]
    days, shifts, req, MAX_DAYS = cfg["days"], cfg["shifts"], cfg["requirement"], cfg["max_days"]

    schedule = {d: {s: [] for s in shifts} for d in days}
    worked, assigned = defaultdict(int), defaultdict(set)

    # pass 1: honor preferences when possible
    for d in days:
        for emp in employees:
            e, pref = emp["name"], emp["prefs"].get(d)
            if pref and len(schedule[d][pref]) < req and worked[e] < MAX_DAYS and d not in assigned[e]:
                schedule[d][pref].append(e)
                assigned[e].add(d)
                worked[e] += 1

    # pass 2: fill minimums without hanging if impossible
    for d in days:
        for s in shifts:
            changed = True
            while len(schedule[d][s]) < req and changed:
                changed = False
                for emp in employees:
                    e = emp["name"]
                    if worked[e] < MAX_DAYS and d not in assigned[e]:
                        schedule[d][s].append(e)
                        assigned[e].add(d)
                        worked[e] += 1
                        changed = True
                        if len(schedule[d][s]) >= req:
                            break
            # if still underfilled here, leave it underfilled and move on

    # print result
    for d in days:
        print(d)
        for s in shifts:
            crew = ", ".join(schedule[d][s]) if schedule[d][s] else "(none)"
            print(f"  {s}: {crew}")
        print()

if __name__ == "__main__":
    main()
